File: go/aistudio/schema.go
package aistudio

import (
	"fmt"
	"sort"
	"strings"
)

var schemaTypes = map[string]int{"string": 1, "number": 2, "integer": 3, "boolean": 4, "array": 5, "object": 6, "null": 7}
var supportedSchemaKeys = map[string]bool{"type": true, "format": true, "description": true, "nullable": true, "enum": true, "items": true, "properties": true, "required": true}

func EncodeSchema(value any) (any, error) {
	schema, ok := value.(map[string]any)
	if !ok {
		return nil, fmt.Errorf("responseSchema must be an object")
	}
	unknown := []string{}
	for key := range schema {
		if !supportedSchemaKeys[key] {
			unknown = append(unknown, key)
		}
	}
	if len(unknown) > 0 {
		sort.Strings(unknown)
		return nil, fmt.Errorf("Unsupported responseSchema fields: %s", strings.Join(unknown, ", "))
	}
	typeName := strings.ToLower(fmt.Sprint(schema["type"]))
	typeEnum, ok := schemaTypes[typeName]
	if !ok {
		return nil, fmt.Errorf("responseSchema.type is required and must be supported")
	}
	encoded := make([]any, 8)
	encoded[0], encoded[1], encoded[2], encoded[3], encoded[4] = typeEnum, schema["format"], schema["description"], schema["nullable"], schema["enum"]
	if raw := schema["items"]; raw != nil {
		item, err := EncodeSchema(raw)
		if err != nil {
			return nil, err
		}
		encoded[5] = item
	}
	if raw := schema["properties"]; raw != nil {
		properties, ok := raw.(map[string]any)
		if !ok {
			return nil, fmt.Errorf("responseSchema.properties must be an object")
		}
		entries := make([]any, 0, len(properties))
		for name, property := range properties {
			item, err := EncodeSchema(property)
			if err != nil {
				return nil, err
			}
			entries = append(entries, []any{name, item})
		}
		encoded[6] = entries
	}
	encoded[7] = schema["required"]
	return trim(encoded), nil
}

File: src/aistudio_client/makersuite/schema.py
"""بخش تأییدشدهٔ JSON Schema را به Schema داخلی MakerSuite تبدیل می‌کند."""

from typing import Any


TYPE_ENUM = {
    "string": 1,
    "number": 2,
    "integer": 3,
    "boolean": 4,
    "array": 5,
    "object": 6,
    "null": 7,
}

SUPPORTED_KEYS = {
    "type", "format", "description", "nullable", "enum",
    "items", "properties", "required",
}


def encode_schema(schema: dict[str, Any]) -> list:
    if not isinstance(schema, dict):
        raise ValueError("responseSchema must be an object")
    unsupported = set(schema) - SUPPORTED_KEYS
    if unsupported:
        names = ", ".join(sorted(unsupported))
        raise ValueError(f"Unsupported responseSchema fields: {names}")

    schema_type = str(schema.get("type", "")).lower()
    if schema_type not in TYPE_ENUM:
        raise ValueError("responseSchema.type is required and must be supported")

    encoded = [None] * 8
    encoded[0] = TYPE_ENUM[schema_type]
    encoded[1] = schema.get("format")
    encoded[2] = schema.get("description")
    encoded[3] = schema.get("nullable")
    encoded[4] = schema.get("enum")
    if schema.get("items") is not None:
        encoded[5] = encode_schema(schema["items"])
    if schema.get("properties") is not None:
        properties = schema["properties"]
        if not isinstance(properties, dict):
            raise ValueError("responseSchema.properties must be an object")
        encoded[6] = [[name, encode_schema(value)] for name, value in properties.items()]
    encoded[7] = schema.get("required")
    return _trim(encoded)


def _trim(values: list) -> list:
    while values and values[-1] is None:
        values.pop()
    return values
